src/Illuminate/Database/Concerns/ManagesTransactions.php

```diff
@@ -38,11 +36,27 @@ public function transaction(Closure $callback, $attempts = 1)
                 $this->handleTransactionException(
                     $e, $currentAttempt, $attempts
                 );
+
+                continue;
             } catch (Throwable $e) {
                 $this->rollBack();
 
                 throw $e;
             }
+
+            // Exceptions on commit need to be handled separately as they automatically roll
+            // back the transaction, resulting in an error if we attempt to manually roll back.
+            try {
+                $this->commit();
```

Comment:
Cosmetic suggestion: We could move `return $callbackReturn;` in here and remove `continue;` from the `catch` block.

Reply:
@staudenmeir I considered that initially, but I think this way makes the pattern of "if something goes wrong, continue; return at the end" more clear. In fact, while working on this PR I initially forgot to put in the first `continue` a few lines up, and it took several minutes to figure out why I was failing unit tests. So while your way is slightly more concise, I think this way is more readable and makes the code easier to modify in the future.